wp/lib/bap_wp/src/compare.ml

```diff
@@ -56,21 +56,45 @@ let init_vars (env1 : Env.t) (env2 : Env.t) (vars : Var.Set.t)
   let inits2, env2 = Pre.init_vars vars env2 in
   (inits1 @ inits2), env1, env2
 
+(** [mk_smtlib2_compare] builds a constraint out of an smtlib2 string that can be used
+    as a comparison predicate between an original and modified binary. *)
+let mk_smtlib2_compare (env1 : Env.t) (env2 : Env.t) (smtlib_str : string) : Constr.t =
+  let var_map1 = (Env.get_var_map env1) in
+  let var_map2 = (Env.get_var_map env2) in
+  let smtlib_str = Env.EnvMap.fold var_map1 ~init:smtlib_str
+      ~f:(fun ~key:var ~data:z3_var smtlib_str ->
+          String.substr_replace_all smtlib_str ~pattern:((Var.name var) ^ "_orig") ~with_:(Z3.Expr.to_string z3_var)
+        ) in
+  let smtlib_str = Env.EnvMap.fold var_map2 ~init:smtlib_str
+      ~f:(fun ~key:var ~data:z3_var smtlib_str ->
+          String.substr_replace_all smtlib_str ~pattern:((Var.name var) ^ "_mod") ~with_:(Z3.Expr.to_string z3_var)
+        ) in
+  info "New smtlib string: %s \n" smtlib_str;
+  let declsym1 = Z3_utils.get_decls_and_symbols env1 in
+  let declsym2 = Z3_utils.get_decls_and_symbols env2 in
+  let declsym = declsym1 @ declsym2 in
+  let ctx = Env.get_context env1 in
+  Z3_utils.mk_smtlib2 ctx smtlib_str declsym
+
 let compare_blocks
     ~input:(input : Var.Set.t)
     ~output:(output : Var.Set.t)
     ~original:(blk1, env1 : Blk.t * Env.t)
     ~modified:(blk2, env2 : Blk.t * Env.t)
+    ~smtlib_post:(smtlib_post : string)
```

Comment:
So what happens if the strings are empty?

Reply:
A good question. I will test this right now.

Reply:
Empty strings get translated to empty clauses `() => ()` which evaluates to true. It seems that we have a few of them in the constraints we are building, but I'm unsure where all of them are being generated. I'm wondering if there is a big performance impact from this.

Reply:
Do we ever simplify the full constraint before passing it to Z3?

Reply:
Oh my, we are not.

Reply:
Not that big a deal. I'll merge this, and we can make this a separate issue.